### PR TITLE
Mulitple fixes for multi-heatmap-trace plots

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/504

which turned out to report two separate issues:

- Off-screen heatmap traces are not correctly updated (broken in https://github.com/plotly/plotly.js/pull/194, fixed in 90d7a94)
- Find closest algo for heatmap traces works again (broken in the lasso PR https://github.com/plotly/plotly.js/pull/154, fixed in 48dc191)

This PR is based off `heatmap-1-xy` of PR https://github.com/plotly/plotly.js/pull/651, to avoid merge conflicts in `heatmap_test.js`. 

This PR should be merged **after** https://github.com/plotly/plotly.js/pull/651.

### CODE REVIEW

File: src/traces/heatmap/plot.js [hunk 1]
```diff
@@ -137,10 +137,24 @@ function plotOne(gd, plotinfo, cd) {
     var imageWidth = Math.round(right - left),
         imageHeight = Math.round(bottom - top);
 
-    // now redraw
+    // setup image nodes
 
     // if image is entirely off-screen, don't even draw it
-    if(imageWidth <= 0 || imageHeight <= 0) return;
```
Comment (etpinard) [inline src/traces/heatmap/plot.js]: `Heatmap.plot` was returning early for off screen traces, but did not clear the exiting nodes.

File: src/traces/heatmap/hover.js [hunk 2]
```diff
@@ -46,8 +48,8 @@ module.exports = function hoverPoints(pointData, xval, yval, hovermode, contour)
             return;
         }
     }
-    else if(Fx.inbox(xval - x[0], xval - x[x.length - 1]) > Fx.MAXDIST ||
-            Fx.inbox(yval - y[0], yval - y[y.length - 1]) > Fx.MAXDIST) {
+    else if(Fx.inbox(xval - x[0], xval - x[x.length - 1]) > MAXDIST ||
+            Fx.inbox(yval - y[0], yval - y[y.length - 1]) > MAXDIST) {
```
Comment (etpinard) [inline src/traces/heatmap/hover.js]: This bug was pretty painful to track down.

It was introduced in https://github.com/plotly/plotly.js/pull/154/commits/d7abd18f92a8c28217ac9fa14a777e5171d99254 when some `graph_interact.js` constants got moved to `plots/cartesian/constants.js`. 

Unfortunately, `Fx.MAXDIST` was still being used in heatmap hover which was `undefined`.

**Reminder:** don't forgot to grep/ack/ag when moving things around!

File: test/image/mocks/heatmap_multi-trace.json [hunk 3]
```diff
@@ -0,0 +1,193 @@
+{
+  "data": [
+    {
+      "x": [
```
Comment (etpinard) [inline test/image/mocks/heatmap_multi-trace.json]: We didn't have a mock for a subplot with multiple heatmap traces on it yet. We have numerous multiple subplots mocks with one heatmap per subplot, but none for multiple heatmap on the same x/y subplot. So, here's one.

File: src/traces/heatmap/calc.js [hunk 4]
```diff
@@ -211,14 +218,17 @@ function makeBoundArray(trace, arrayIn, v0In, dvIn, numbricks, ax) {
     }
     else {
         dv = dvIn || 1;
-        if(v0In === undefined) v0 = 0;
+
+        if(Array.isArray(arrayIn) && arrayIn.length === 1) v0 = arrayIn[0];
+        else if(v0In === undefined) v0 = v0In;
```
Comment (mdtusz) [inline src/traces/heatmap/calc.js]: Is this a mistake?

If the condition is met, we'll be adding `undefined + dv * i` below which will `NaN`.

Comment (etpinard) [inline src/traces/heatmap/calc.js]: Good :eyes: should be `if(v0In === undefined) v0 = 0;` like the line I removed. Thanks!

**Comment by mdtusz** [comment]
💃

File: src/traces/heatmap/hover.js (hunk 2 above)
Comment (alexcjohnson) [inline src/traces/heatmap/hover.js]: sorry about that @etpinard - as a bit of karmic penance, I fixed this myself in #673 then had to undo my fix when I merged master and got your fix too!

**Pull request merged**